### DISCUSSION TITLE
Version 5.0.0 alpha 2 available

### DISCUSSION THREAD

**Discussion by mkleehammer** [mkleehammer/pyodbc#1254] [Announcements]
I've just merged the 5.0 branch into master and `uploaded` 5.0.0a2 to PyPI for testing.  I created a 4.0 branch for security updates.

We were finally able to drop Python 2 support!  This supports Python 3.7-3.12.

To support 3.12, Unicode handling had to be reworked.  Unfortunately I've was not able to port the `fastexecmany` code yet.  Ideally this feature should use the same parameter encoding code as the rest of pyodbc, so I'm planning on rewriting it.  Since the release of 3.12 is coming up, I might have to release 5.0 without it and follow up with a 5.1 release that improves performance.  I'm also thinking the feature should be renamed to something that better describes what it does like "array binding" or something.  Any thoughts or suggestions?

I've ported the unit tests to use [pytest](https://pytest.org).  I've put it, plus flake8 and pylint, into requirements-dev.txt.

To test locally, obviously you can build and install, but you can also test without installing by:

- Build the library into the root using `python setup.py build_ext -i`
- Use this library by running pytest from the root: `pytest -v tests/*_test.py`

I ported the 3 unit tests for databases supported by Github Actions & AppVeyor: SQL Server, PostgreSQL, and MySQL.  If you use one of the other databases we had tests for and want to port them to pytest, it would be greatly appreciated.

Please give it a spin and let me know if you find any problems.

**Comment by keitherskine** [comment]
Excellent news, @mkleehammer !  So glad Python 2.7 has been dropped, and as a bonus, the codebase now works with 3.12 as well.

**Comment by mkleehammer** [comment]
BTW, I want to thank @keitherskine for initially setting up the Github & AppVeyor CI.  That's been a huge productivity boost and made this release much easier.

**Comment by mkleehammer** [comment]
Note to future readers: python setup.py build_ext --inplace will not work now that there is a pyproject.toml file.  You can confirm this by renaming the toml file and running the command.

I will look into this in the future, but it looks like I'll have to dig into the guts of setuptools.
